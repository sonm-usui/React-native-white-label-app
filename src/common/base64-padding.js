// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
