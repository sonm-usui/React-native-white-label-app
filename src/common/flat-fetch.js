// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
