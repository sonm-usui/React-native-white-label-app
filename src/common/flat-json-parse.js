// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
