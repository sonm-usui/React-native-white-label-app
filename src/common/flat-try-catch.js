// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
