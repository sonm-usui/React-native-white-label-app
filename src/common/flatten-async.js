// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
