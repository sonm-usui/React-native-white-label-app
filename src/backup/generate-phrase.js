// @flow

import React, { Component } from 'react'
import { Image } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { headerDefaultOptions } from '../navigation/navigation-header-config'
import type { Store } from '../store/type-store'
import { Container, CustomView, CustomText, Loader } from '../components'
import { colors, fontSizes } from '../common/styles'
import {
  genRecoveryPhraseRoute,
  verifyRecoveryPhraseRoute,
  settingsRoute,
} from '../common'
import { color } from '../common/styles'
import type {
  GenerateRecoveryPhraseProps,
  GenerateRecoveryPhraseState,
  Passphrase,
  ChatBubbleDimensions,
  PassphraseTextProps,
  PassphraseErrorProps,
} from './type-backup'
import { BACKUP_STORE_STATUS } from './type-backup'
import { generateRecoveryPhrase } from './backup-store'
import {
  SHOW_RECOVERY_PHRASE_TEST_ID,
  SUBMIT_RECOVERY_PHRASE_TEST_ID,
  SUBMIT_RECOVERY_PHRASE_BUTTON_TITLE,
} from './backup-constants'
import styles, { chatBubbleTextOffset } from './styles'
import { getBackupPassphrase, getBackupStatus } from '../store/store-selector'
import { PASSPHRASE_GENERATION_ERROR } from '../common'
import { withStatusBar } from '../components/status-bar/status-bar'
import { Button } from '../components/buttons/button'

const transparentBands = require('../images/transparentBands.png')
const textBubble = require('../images/textBubble.png')

const PassphraseLoader = () => {
  return (
    <CustomView style={[styles.genRecoveryPhraseLoadingContainer]}>
      <Loader delay={10} showMessage={false} />
    </CustomView>
  )
}

const PassphraseError = (props: PassphraseErrorProps) => {
  return (
    <CustomView
      style={[
        styles.genRecoveryPhraseContainer,
        {
          width:
            props.chatBubbleDimensions && props.chatBubbleDimensions.width
              ? props.chatBubbleDimensions.width - chatBubbleTextOffset
              : null,
        },
      ]}
    >
      <CustomText transparentBg darkgray center>
        {PASSPHRASE_GENERATION_ERROR}
      </CustomText>
    </CustomView>
  )
}

const PassphraseText = (props: PassphraseTextProps) => {
  return (
    <CustomView
      style={[
        styles.genRecoveryPhraseContainer,
        {
          //width of textContainer is chatBubbleDimensions width minus chatBubbleTextOffset
          //so that passPhrase text is not cutting beyond the chatBubble
          width:
            props.chatBubbleDimensions && props.chatBubbleDimensions.width
              ? props.chatBubbleDimensions.width - chatBubbleTextOffset
              : null,
        },
      ]}
    >
      <CustomText
        transparentBg
        style={[styles.genRecoveryPhrase]}
        testID={SHOW_RECOVERY_PHRASE_TEST_ID}
      >
        {props.recoveryPassphrase.phrase}
      </CustomText>
    </CustomView>
  )
}

export class GenerateRecoveryPhrase extends Component<
  GenerateRecoveryPhraseProps,
  GenerateRecoveryPhraseState
> {
  state = {
    chatBubbleDimensions: {
      width: 0,
      height: 0,
      x: 0,
      y: 0,
    },
  }

  componentDidMount() {
    this.props.generateRecoveryPhrase()
  }

  verifyRecoveryPhrase = () => {
    const {
      navigation: { navigate },
      route,
    } = this.props

    navigate(verifyRecoveryPhraseRoute, {
      recoveryPassphrase: this.props.recoveryPassphrase.hash,
      initialRoute: route.params.initialRoute,
    })
  }
  backToSettings = () => {
    const {
      navigation: { navigate },
    } = this.props
    navigate(settingsRoute)
  }
  onRecoveryPhraseGoBack = () => {
    const {
      navigation: { goBack },
    } = this.props
    goBack(null)
  }

  //TODO fix refactor UI
  //first the image should be used as a ImageBackground component
  ImageContents = (
    recoveryStatus: string,
    recoveryPassphrase: Passphrase,
    chatBubbleDimensions: ChatBubbleDimensions
  ) => {
    if (recoveryStatus === BACKUP_STORE_STATUS.GENERATE_PHRASE_LOADING) {
      return <PassphraseLoader />
    }

    if (
      recoveryStatus === BACKUP_STORE_STATUS.GENERATE_PHRASE_FAILURE ||
      recoveryStatus === BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_FAILURE
    ) {
      // This block is where we need to try handling passphrase generation differently
      return <PassphraseError chatBubbleDimensions={chatBubbleDimensions} />
    }
    return (
      <PassphraseText
        recoveryPassphrase={recoveryPassphrase}
        chatBubbleDimensions={chatBubbleDimensions}
      />
    )
  }

  setChatBubbleDimensions = (chatBubbleDimensions: ChatBubbleDimensions) => {
    this.setState({
      chatBubbleDimensions,
    })
  }

  render() {
    // for viewRecover mode
    const viewOnlyMode = this.props.route.params.viewOnlyMode ? true : false
    const { recoveryPassphrase, recoveryStatus } = this.props
    const disableButton =
      recoveryStatus === BACKUP_STORE_STATUS.GENERATE_PHRASE_FAILURE ||
      recoveryStatus === BACKUP_STORE_STATUS.GENERATE_PHRASE_LOADING ||
      recoveryStatus === BACKUP_STORE_STATUS.GENERATE_BACKUP_FILE_FAILURE

    return (
      <Container style={[styles.genRecovery]} safeArea>
        <Image source={transparentBands} style={[styles.backgroundImage]} />
        <Container>
          <CustomView>
            <CustomText
              transparentBg
              numberOfLines={1}
              style={[styles.genRecoveryText]}
            >
              {viewOnlyMode
                ? 'Your Recovery Phrase'
                : 'Recovery Phrase generated'}
            </CustomText>
          </CustomView>
          <CustomView center>
            <CustomText
              center
              transparentBg
              style={[styles.genRecoveryMessage]}
            >
              This is your Recovery Phrase. Write it down, and don't share it
              with anyone.
            </CustomText>
          </CustomView>
          <CustomView center>
            <Image
              source={textBubble}
              style={[styles.imageIcon]}
              onLayout={(event) =>
                this.setChatBubbleDimensions(event.nativeEvent.layout)
              }
            />
            {this.ImageContents(
              recoveryStatus,
              recoveryPassphrase,
              this.state.chatBubbleDimensions
            )}
          </CustomView>
          <CustomView center>
            <CustomText
              center
              transparentBg
              style={[styles.genRecoverySecondMessage]}
            >
              You will need this Recovery Phrase to restore from a backup. Keep
              it safe.
            </CustomText>
          </CustomView>
        </Container>

        {viewOnlyMode && (
          <CustomView>
            <CustomView center />
            <Button
              buttonStyle={[styles.submitButton]}
              label={'Got It'}
              onPress={this.backToSettings}
              testID={SUBMIT_RECOVERY_PHRASE_TEST_ID}
              labelStyle={{
                color: color.bg.eleventh.color,
                fontWeight: '600',
                fontSize: fontSizes.size3,
              }}
            />
          </CustomView>
        )}
        {!viewOnlyMode && (
          <CustomView>
            <CustomView center>
              <CustomText
                transparentBg
                center
                style={[styles.genRecoverySmallMessage]}
              >
                Are you sure you wrote it down?
              </CustomText>
            </CustomView>
            <Button
              onPress={this.verifyRecoveryPhrase}
              label={SUBMIT_RECOVERY_PHRASE_BUTTON_TITLE}
              disabled={disableButton}
              buttonStyle={styles.submitButton}
              labelStyle={{
                color: color.bg.eleventh.color,
                fontWeight: '600',
                fontSize: 18,
              }}
              testID={SUBMIT_RECOVERY_PHRASE_TEST_ID}
            />
          </CustomView>
        )}
      </Container>
    )
  }
}

const mapStateToProps = (state: Store) => {
  return {
    recoveryPassphrase: getBackupPassphrase(state),
    recoveryStatus: getBackupStatus(state),
  }
}
const mapDispatchToProps = (dispatch) =>
  bindActionCreators({ generateRecoveryPhrase }, dispatch)

export const generateRecoveryPhraseScreen = {
  routeName: genRecoveryPhraseRoute,
  screen: withStatusBar({ color: color.bg.eleventh.color })(
    connect(mapStateToProps, mapDispatchToProps)(GenerateRecoveryPhrase)
  ),
  options: headerDefaultOptions({
    headline: undefined,
    headerHideShadow: true,
    transparent: true,
    headerStyles: {
      backgroundColor: color.bg.eleventh.color
    },
    arrowColor: colors.white,
  })
}
