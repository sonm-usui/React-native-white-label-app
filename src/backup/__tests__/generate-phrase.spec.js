// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { GenerateRecoveryPhrase } from '../generate-phrase'
import { settingsRoute } from '../../common'
import { BACKUP_STORE_STATUS } from '../type-backup'

describe('<GenerateRecoveryPhrase />', () => {
  const recoveryPassphrase = 'hello some passphrase'

  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
  })

  const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    setParams: jest.fn(),
  }

  const route = {
    params: {
      recoveryPassphrase,
      initialRoute: settingsRoute,
      navigateBack: jest.fn(),
    },
  }

  it('should match snapshot', () => {
    const tree = renderer.create(
      <GenerateRecoveryPhrase
        recoveryPassphrase={{
          phrase: recoveryPassphrase,
          salt: 'salt',
          hash: 'hash',
        }}
        generateRecoveryPhrase={jest.fn()}
        hydrateCloudBackup={jest.fn()}
        recoveryStatus={BACKUP_STORE_STATUS.GENERATE_PHRASE_SUCCESS}
        navigation={navigation}
        route={route}
      />
    )
    tree.getInstance().setState({
      chatBubbleDimensions: {
        width: 360,
        height: 199,
        x: 0,
        y: 0,
      },
    })
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('should match snapshot for loading state', () => {
    const tree = renderer.create(
      <GenerateRecoveryPhrase
        recoveryPassphrase={{
          phrase: recoveryPassphrase,
          salt: 'salt',
          hash: 'hash',
        }}
        generateRecoveryPhrase={jest.fn()}
        hydrateCloudBackup={jest.fn()}
        recoveryStatus={BACKUP_STORE_STATUS.GENERATE_PHRASE_LOADING}
        navigation={navigation}
        route={route}
      />
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('should match snapshot for error state', () => {
    const tree = renderer.create(
      <GenerateRecoveryPhrase
        recoveryPassphrase={{
          phrase: recoveryPassphrase,
          salt: 'salt',
          hash: 'hash',
        }}
        generateRecoveryPhrase={jest.fn()}
        hydrateCloudBackup={jest.fn()}
        recoveryStatus={BACKUP_STORE_STATUS.GENERATE_PHRASE_FAILURE}
        navigation={navigation}
        route={route}
      />
    )
    tree.getInstance().setState({
      chatBubbleDimensions: {
        width: 360,
        height: 199,
        x: 0,
        y: 0,
      },
    })

    expect(tree.toJSON()).toMatchSnapshot()
  })
})
