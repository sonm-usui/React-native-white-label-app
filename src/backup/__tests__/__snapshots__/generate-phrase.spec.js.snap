// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenerateRecoveryPhrase /> should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  safeArea={true}
  style={
    Object {
      "backgroundColor": "#4A90E2",
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "default": "RelativeImageStub",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": "2%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Recovery Phrase generated
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 20,
              "marginBottom": 5,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        This is your Recovery Phrase. Write it down, and don't share it with anyone.
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Image
        onLayout={[Function]}
        source={
          Object {
            "default": "RelativeImageStub",
          }
        }
        style={
          Array [
            Object {
              "marginBottom": 0,
              "resizeMode": "stretch",
              "width": "100%",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "left": 20,
            "paddingHorizontal": "7%",
            "position": "absolute",
            "top": 20,
            "width": 320,
          }
        }
      >
        <Text
          accessibilityLabel="hello some passphrase"
          accessible={true}
          allowFontScaling={false}
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "fontFamily": "System",
              },
              Object {
                "backgroundColor": "#333333",
                "color": "#ffffff",
              },
              null,
              null,
              Object {
                "backgroundColor": "transparent",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              Object {
                "color": "#4A90E2",
                "flexWrap": "wrap",
                "fontSize": 18,
                "fontWeight": "900",
                "lineHeight": 28,
                "textAlign": "center",
              },
            ]
          }
          suppressHighlighting={true}
          testID="show-recovery-passphrase"
        >
          hello some passphrase
        </Text>
      </View>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "bold",
              "lineHeight": 20,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        You will need this Recovery Phrase to restore from a backup. Keep it safe.
      </Text>
    </View>
  </View>
  <View
    style={Object {}}
  >
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 17,
              "marginBottom": "3%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Are you sure you wrote it down?
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 5,
          "elevation": 7,
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
          "marginBottom": 15,
          "marginLeft": "2%",
          "marginRight": "2%",
          "opacity": 1,
          "shadowColor": "#777777",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 5,
          "width": "94%",
        }
      }
      testID="submit-recovery-passphrase"
    >
      <Text
        style={
          Object {
            "color": "#4A90E2",
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Continue
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`<GenerateRecoveryPhrase /> should match snapshot for error state 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  safeArea={true}
  style={
    Object {
      "backgroundColor": "#4A90E2",
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "default": "RelativeImageStub",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": "2%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Recovery Phrase generated
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 20,
              "marginBottom": 5,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        This is your Recovery Phrase. Write it down, and don't share it with anyone.
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Image
        onLayout={[Function]}
        source={
          Object {
            "default": "RelativeImageStub",
          }
        }
        style={
          Array [
            Object {
              "marginBottom": 0,
              "resizeMode": "stretch",
              "width": "100%",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "left": 20,
            "paddingHorizontal": "7%",
            "position": "absolute",
            "top": 20,
            "width": 320,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "fontFamily": "System",
              },
              Object {
                "backgroundColor": "#333333",
                "color": "#ffffff",
              },
              null,
              Object {
                "textAlign": "center",
              },
              Object {
                "backgroundColor": "transparent",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              Object {
                "color": "#4A4A4A",
              },
              null,
              null,
            ]
          }
          suppressHighlighting={true}
        >
          Uh oh... Looks like something went wrong.
        </Text>
      </View>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "bold",
              "lineHeight": 20,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        You will need this Recovery Phrase to restore from a backup. Keep it safe.
      </Text>
    </View>
  </View>
  <View
    style={Object {}}
  >
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 17,
              "marginBottom": "3%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Are you sure you wrote it down?
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 5,
          "elevation": 7,
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
          "marginBottom": 15,
          "marginLeft": "2%",
          "marginRight": "2%",
          "opacity": 1,
          "shadowColor": "#777777",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 5,
          "width": "94%",
        }
      }
      testID="submit-recovery-passphrase"
    >
      <Text
        style={
          Object {
            "color": "#4A90E2",
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Continue
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`<GenerateRecoveryPhrase /> should match snapshot for loading state 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  safeArea={true}
  style={
    Object {
      "backgroundColor": "#4A90E2",
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "default": "RelativeImageStub",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 20,
              "fontWeight": "600",
              "marginBottom": "2%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Recovery Phrase generated
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 20,
              "marginBottom": 5,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        This is your Recovery Phrase. Write it down, and don't share it with anyone.
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Image
        onLayout={[Function]}
        source={
          Object {
            "default": "RelativeImageStub",
          }
        }
        style={
          Array [
            Object {
              "marginBottom": 0,
              "resizeMode": "stretch",
              "width": "100%",
            },
          ]
        }
      />
      <View
        style={
          Object {
            "padding": 40,
            "position": "absolute",
            "top": 0,
          }
        }
      />
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "alignSelf": "center",
              "fontSize": 16,
              "fontWeight": "bold",
              "lineHeight": 20,
              "paddingHorizontal": 20,
            },
          ]
        }
        suppressHighlighting={true}
      >
        You will need this Recovery Phrase to restore from a backup. Keep it safe.
      </Text>
    </View>
  </View>
  <View
    style={Object {}}
  >
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "fontSize": 14,
              "fontWeight": "bold",
              "lineHeight": 17,
              "marginBottom": "3%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Are you sure you wrote it down?
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 5,
          "elevation": 7,
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
          "marginBottom": 15,
          "marginLeft": "2%",
          "marginRight": "2%",
          "opacity": 1,
          "shadowColor": "#777777",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 5,
          "width": "94%",
        }
      }
      testID="submit-recovery-passphrase"
    >
      <Text
        style={
          Object {
            "color": "#4A90E2",
            "fontSize": 18,
            "fontWeight": "600",
          }
        }
      >
        Continue
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
